### ISSUE TITLE
Wasm modules not instantiating correctly with shared memory

### ISSUE THREAD

**Issue by WireWhiz** [rust-lang/rust#116050]
Hello, I'm currently attempting to use rust built to wasm as a scripting language.  The exact issue is that the first memory allocation or use of memory from that allocation in a wasm function causes a panic. This is likely because I'm loading several different wasm modules with the same imported shared memory and them not interacting quite right to instantiate that, or something similar. 

## Repo steps

I'm unsure the level of complexity needed to repro the issue so I'm just providing source here to be safe. (Just poc/research stuff so it's pretty messy)

To do this I first build my "scripts" with a shell script:

[testScript1.rs](https://github.com/rust-lang/rust/files/12698421/testScript1.txt)
[testScript2.rs](https://github.com/rust-lang/rust/files/12698427/testScript2.txt)

```shell
export RUSTFLAGS="--cfg=web_sys_unstable_apis -C target-feature=+atomics,+bulk-memory,+mutable-globals -C link-arg=--no-entry -C link-arg=--shared-memory -C link-arg=--import-memory -C link-arg=--export-memory -C link-arg=--max-memory=2147483648"
echo "Compiling scripts with flags:" $RUSTFLAGS
# should be noted I swap between wasm32-unknown-unknown and this target for testing and the bug occurs in both
cargo +nightly build --target wasm32-wasi-preview1-threads -Z build-std=std,panic_abort
```

I then load and instantiate all the compiled wasm modules using wasmtime:
[main.rs](https://github.com/rust-lang/rust/files/12698431/main.txt)

## Exact conditions
The exact conditions that seem to trigger this are:
1. shared and imported memory
2. at least 2 module instances, both must be different code. Error is not triggered if the instances are from the same module.
3. wasm target does not matter, currently my setup is using the threads target since I wanted to test that, but I haven't gotten that far. Error still triggers on wasm32-unknown-unknown.

## Related
When I first discovered the error I was attempting to figure out threading as it seemed that I had the whole multiple modules setup working pretty solid, though I found out that I was instantiating SharedMemory the wrong way (even though modules successfully loaded and functioned) and once I switched to the correct method this error started with no other changes. 
This is all documented in more detail here: https://github.com/bytecodealliance/wasmtime/issues/7054

## Does the wasm spec cover this?
I suspect that I may be attempting a use case that hasn't really been covered yet so the spec may not cover it, or it might not be possible within spec and I've edge cased myself into getting it working up until now. So a side tangent is how could I go about reporting/contributing to that if necessary.

**Comment by g0djan** [comment]
Hey, thanks for raising the issue. I'll take a look at it next week

**Comment by g0djan** [comment]
@WireWhiz Hey, can you provide more details on how to reproduce the issue? A step by step manual would be great but answers to next questions might be enough

1. Do I compile each testSricpt.rs to wasm32-wasi-preview1-threads separately? Cargo complains on missing main function in that case, but maybe I missed its libraries
2. What target main.rs should be compiled to? I thought it should be compiled to wasm32-wasi-preview1-threads as well and should be run on wasmtime, but it imports wasmtime_wasi so I'm not sure

**Comment by WireWhiz** [comment]
@g0djan Sorry for taking so long to respond. Yes, each test script is compiled to a different wasm binary. The way I have it currently set up in my project is that they're both separate crates in one workspace so a single cargo build command will build both. There is also no main function, as build them with the --no-entry flag set in RUSTFLAGS using that shell script. 

Main.rs is compiled to a native executable and loads the scripts using wasmtime as a library. 

I realize now I haven't provided a fair about of other files that I referenced here, I should probably have created a slimmed down example.

**Comment by g0djan** [comment]
Took me longer to reply after though, I've been on holidays.
> I should probably have created a slimmed down example.
If you can share a repo with build commands it would be ideal.

Is this problem blocking you atm?

**Comment by WireWhiz** [comment]
It is blocking me, but it's a side project so not hyper urgent.

I'm currently working in this repo on the wasmScriptingTests branch.
[https://github.com/WireWhiz/BraneEngine](https://github.com/WireWhiz/BraneEngine)

All the crates are under `src/common/scripting`. You'll want to run `compileScript.sh` in `TestWasmScripts` to generate the two .wasm files, then manually copy those over to WasmtimeTest (though I think I committed some already built/copied ones over so that may not be necessary) and then cargo run WasmtimeTest.

I'll check tonight to make sure the current push builds and repos the issue.

If there's anything I can help with lmk, I originally was working on a scripting language for this engine so I have some llvm experience and I wouldn't mind learning how to contribute here. I was thinking about reviving that scripting language and porting it to build to wasm, but I'd basically to turn it into rust + the multi-memory proposal at this point.

**Comment by g0djan** [comment]
Okay, I followed your steps.

Do you get

```
Module load time: 2.728532667s
Module load time: 2.666379208s
Instance1 instantiation time: 377.5µs
Instance2 instantiation time: 42ns
thread 'main' panicked at src/main.rs:158:92:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
?

**Comment by WireWhiz** [comment]
Yeah, I'm getting that, I ended up never getting around to making sure the branch was compiling. I'm pretty busy but I'm working on making a slimmed down repo of the error. I'll update here when that's ready so I can stop wasting your time.

**Comment by g0djan** [comment]
Okay, thanks for working on a slimmed down repo

**Comment by WireWhiz** [comment]
I've created a much smaller sandbox repo:
https://github.com/WireWhiz/WasmTests

I'm getting a different bug than I was before interestingly enough, when instantiating a second module, whether in the same store or not, it seems to hang on an async operation in the wasm memory initialization function that's trying to access the shared memory. I tried out instantiating two module1 instances to see if it would trigger the same bug but two of the same module worked just fine, and then the bug would still trigger on module2 instantiating.  Reversing the order of instantiation (module2, and then module1) works just fine and all the functions run perfectly.

My guess is that instantiating two different binaries with the same shared memory is undefined behavior at the moment? I seem to get a different bug every time I slightly modify the project.

**Comment by WireWhiz** [comment]
Hmm, that might be something to do with my environment because I built those with the wasm32-unknown-unknown target. I'll try updating and see if that fixes it.

This is a different bug than the memory alloc one so I'm not sure if it still happens on wasi, but if the bug persists after I update toolchains I'll update the text executable to be able to support wasi as well

**Comment by WireWhiz** [comment]
Ok after running rustup update everything seems to work on my end. I also added in a toggle to initialize the runtime both with and without wasi and it seems to work both ways. 

To be thorough I pasted in some expanded macros from the source project I was triggering the original bug in and it seemed to continue working. 

This leads me to believe whatever I was triggering was already fixed in the time it took me to get around to re-testing stuff, or it was an implementation error on my part, so I'm going to close this issue for now and reopen if it pops up again with hopefully more info.
